### PR TITLE
octave: use qt-5 mkDerivation for wrapQtAppsHook

### LINKED ISSUE #96984: Octave 5.2.0: error finding Qt platform plugin on Darwin
**Describe the bug**
When running Octave on Darwin I observed the following error:

```
qt.qpa.plugin: Could not find the Qt platform plugin "cocoa" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```

**To Reproduce**
1. Build `octaveFull` on Darwin (required #96979 and #96982 on top of nixpkgs master for me)
2. Octave will abort.  Run octave with `OCTAVE_SHOW_GUI_MESSAGES=1` (took me ages to find this!) and you should see the error message above.

**Expected behavior**
Octave should pick up the plugin.

I found I could work around the issue by setting `QT_PA_PLATFORM_PLUGIN_PATH=/nix/store/...-qtbase-5.12.7-bin/lib/qt-5.12.7/plugins/platforms`, so one potential fix would be to wrap octave to set this.  I'll make a PR for that.

**Notify maintainers**

@idontgetoutmuch @doronbehar

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
 - system: `"x86_64-darwin"`
 - host os: `Darwin 19.5.0, macOS `
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.6`
 - channels(...): `"nixpkgs-20.09pre239985.a332da8588a"`
 - nixpkgs: `/Users/.../.nix-defexpr/channels/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by graham33**
###### Motivation for this change

This is to ensure QT_QPA_PLATFORM_PLUGIN_PATH is set on Darwin.

Fixes #96984

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by worldofpeace — CHANGES_REQUESTED** [review]
`wrapQtAppsHook` exists to do this for us. If you change the deriver from `stdenv.mkDerivation` to `mkDerivation` it will add the hook automatically. Also note, while those seem to be named exactly the same, `mkDerivation` will come from the `qt5` attribute set because of the callPackage function (that I believe) this package should be using. You should also condition it's adding on `optionals enableQt`.

**Comment by doronbehar** [comment]
> wrapQtAppsHook exists to do this for us. If you change the deriver from stdenv.mkDerivation to mkDerivation it will add the hook automatically. Also note, while those seem to be named exactly the same, mkDerivation will come from the qt5 attribute set because of the callPackage function (that I believe) this package should be using. You should also condition it's adding on optionals enableQt.

Correct. But, since octave is an interpreter, and it's build is so heavy, I'd argue we need an external wrapper derivation, so we would also be able to have a `octave.withPackages` interface, see https://github.com/NixOS/nixpkgs/issues/65398 and [this comment](https://github.com/NixOS/nixpkgs/issues/86284#issuecomment-622353703).

I thought I would be able to postpone this until the 6.0.0 release, which I can't wait to get a hold on for a long time now.

Something modeled after `wrapNeovim` and `wrapMpv` could be done (see https://github.com/NixOS/nixpkgs/pull/88620/files). This change should also start with this:

```diff
diff --git i/pkgs/development/interpreters/octave/default.nix w/pkgs/development/interpreters/octave/default.nix
index 06d69ed8117..e4b527eaa6e 100644
--- i/pkgs/development/interpreters/octave/default.nix
+++ w/pkgs/development/interpreters/octave/default.nix
@@ -148,6 +148,7 @@ stdenv.mkDerivation rec {
   passthru = {
     inherit version;
     sitePath = "share/octave/${version}/site";
+    inherit enableQt enableJIT enableJava enableReadline;
   };
 
   meta = {
```

**Comment by doronbehar** [comment]
Maybe double wrapping wouldn't be that bad. I'm experimenting a fix for this on top of octave 6's release candidate.

**Comment by doronbehar** [comment]
P.S on https://github.com/doronbehar/nixpkgs/blob/pkg/octave/pkgs/development/interpreters/octave/default.nix

**Comment by graham33** [comment]
@doronbehar that makes sense, thanks.  I'll update this PR to just use the right mkDerivation and test that, so we can either choose to merge or wait for a bigger refactoring.

**Comment by doronbehar** [comment]
@graham33 please, if you'd test this on Darwin it'll be helpful. Something like this should work:

https://github.com/doronbehar/nixpkgs/commit/7c6db8863e70add4d9cb63772d2584584b5a94b8

The refactoring I'm working on are for Octave 6, and they come along other changes, so don't wait for me. I also did not find time to test the `mkDerivation` thing while testing my changes.

**Comment by graham33** [comment]
Updated but not yet tested.

**Comment by graham33** [comment]
Result of `nixpkgs-review pr 96985` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>octaveFull</li>
  </ul>
</details>

**Comment by graham33** [comment]
I've tested this briefly on Darwin (also on NixOS) and I get a working Octave.  For Darwin I had to cherry-pick 8e4e38cdcdc5f2be6875e6506d0789e9e5411ea6 from staging to get a working build (it hasn't made it to master yet).

Just to briefly summarise the state of Octave on Darwin (at least from my brief testing, I don't use Darwin day-to-day): with the above fix the GUI works as long as you run with `octave --force-gui`, and as I mentioned basic editing/calculations/plotting seem to work.  There are issues with the system menus sometimes not working, and it tends to hang/crash on exit.  I imagine these issues might be improved via a Qt upgrade, but I haven't tried that.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -10123,14 +10123,16 @@ in
...
     jdk = jdk8; # TODO: remove override https://github.com/NixOS/nixpkgs/pull/89731
   };
   octave-jit = callPackage ../development/interpreters/octave {
     python = python3;
     enableJIT = true;
+    mkDerivation = stdenv.mkDerivation;
     jdk = jdk8; # TODO: remove override https://github.com/NixOS/nixpkgs/pull/89731
   };
-  octaveFull = (lowPrio (libsForQt512.callPackage ../development/interpreters/octave {
+  octaveFull = (lowPrio (libsForQt5.callPackage ../development/interpreters/octave {
```
Comment (doronbehar) [inline pkgs/top-level/all-packages.nix:10135]: I'd remove the `lowPrio` along with this change. It's a leftover from quiet some time ago.

**Review by doronbehar — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (graham33) [reply pkgs/top-level/all-packages.nix]: Done (not really sure if there's anything to be tested, let me know if so).

**Comment by doronbehar** [comment]
Thanks :)

**Pull request merged**